### PR TITLE
Support JSON2CSV and CSV2JSON output format conversion

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Support JSON to CSV and CSV to JSON output format conversion

## Motivation and Context
This PR implements one of the pending items in issue #6286
in S3 API a user can request CSV output for a JSON document
and a JSON output for a CSV document. This PR refactors
the code a little bit to bring this feature.

## Regression
No

## How Has This Been Tested?
Using some custom changes to `mc`

```diff
git diff
diff --git a/cmd/client-s3.go b/cmd/client-s3.go
index 2d1634f4..d670875c 100644
--- a/cmd/client-s3.go
+++ b/cmd/client-s3.go
@@ -425,10 +425,16 @@ func (c *s3Client) Select(expression, sseKey string) (io.ReadCloser, *probe.Erro
                        },
                }
                opts.OutputSerialization = minio.SelectObjectOutputSerialization{
-                       JSON: &minio.JSONOutputOptions{
+                       CSV: &minio.CSVOutputOptions{
                                RecordDelimiter: "\n",
+                               FieldDelimiter:  ",",
                        },
                }
+               // opts.OutputSerialization = minio.SelectObjectOutputSerialization{
+               //      JSON: &minio.JSONOutputOptions{
+               //              RecordDelimiter: "\n",
+               //      },
+               // }
        }
        if sseKey != "" {
                key, err := encrypt.NewSSEC([]byte(sseKey))
```

This converts JSON document to CSV output.

```
~ mc sql --query "select * from S3Object" myminio/sjm-airlines/people.json
Michael,31
Andy,30
Justin,19
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#5cf69395-fb81-4d3e-a469-7acaf80144fb) |
<a name="5cf69395-fb81-4d3e-a469-7acaf80144fb"></a>
### 6910-32d7e87/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30626
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 126aa3fdce41:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 46 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 6 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 41 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 51 seconds
(8/13) Running minio-go tests ... done in 58 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 22 seconds
(10/13) Running minio-js tests ... FAILED in 33 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-4d6794cd-039f-4a70-8cfc-aa862cc85a04, events: s3:ObjectCreated:*",
  "duration": 20074,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#b49640d1-8e8c-414a-b6c8-b3bd1fba8a8c) |
<a name="b49640d1-8e8c-414a-b6c8-b3bd1fba8a8c"></a>
### 6910-8d65ee8/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30764
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fa6468d919c9:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 57 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 36 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 51 seconds
(8/13) Running minio-go tests ... done in 45 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 25 seconds
(10/13) Running minio-js tests ... FAILED in 29 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-28c12208-e133-428f-afa0-f78185c7fc70, events: s3:ObjectCreated:*",
  "duration": 20250,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by sinhaashish** [comment]
The sample json as shown below  is failing when queried for csv output.
`{	"Id": 1,
	"Name": "Anshu",
	"Address": "Templestowe",
	"Car": "Jeep"
}
{
	"Id": 2,
	"Name": "Ben Mostafa",
	"Address": "Las Vegas",
	"Car": "Mustang"
}
{
	"Id": 3,
	"Name": "Rohan Wood",
	"Address": "Wooddon",
	"Car": "VW"
}
{
	"Id": 4,
	"Name": "Anshu",
	"Address": "Templestowe",
	"Car": "Jeep"
}`

**Comment by harshavardhana** [comment]
> The sample json as shown below is failing when queried for csv output.
> `{	"Id": 1, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep" } { "Id": 2, "Name": "Ben Mostafa", "Address": "Las Vegas", "Car": "Mustang" } { "Id": 3, "Name": "Rohan Wood", "Address": "Wooddon", "Car": "VW" } { "Id": 4, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep" }`

This is an incorrect JSON @sinhaashish we do not support JSON document, you need to provide a multiline JSON. Of course the harder problem is detecting that its not a multiline JSON so it is not returning anything. 

Such as this
```json
{"Id": 1, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep"}
{"Id": 2, "Name": "Ben Mostafa", "Address": "Las Vegas", "Car": "Mustang"}
{"Id": 3, "Name": "Rohan Wood", "Address": "Wooddon", "Car": "VW"}
{"Id": 4, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep" }
```

```
~ mc sql -e "select * from s3object" myminio/sjm-airlines/test.json
1,Anshu,Templestowe,Jeep
2,Ben Mostafa,Las Vegas,Mustang
3,Rohan Wood,Wooddon,VW
4,Anshu,Templestowe,Jeep
```

And for population.csv
```
~ mc sql -e "select * from s3object" myminio/sjm-airlines/population.csv | jq .
{
  "PopTotal": "1291.622",
  "PopFemale": "661.578",
  "PopMale": "630.044",
  "AgeGrpSpan": "5",
  "AgeGrpStart": "0",
  "AgeGrp": "0-4",
  "MidPeriod": "1950.5",
  "Time": "1950",
  "Variant": "High",
  "VarID": "3",
  "Location": "Afghanistan",
  "LocID": "4"
}
...
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#d82b72a5-d04b-4558-9da9-a1b911538b10) |
<a name="d82b72a5-d04b-4558-9da9-a1b911538b10"></a>
### 6910-376c73a/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30743
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 26ecf8b464f0:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 42 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 2 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 32 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 42 seconds
(8/13) Running minio-go tests ... done in 23 seconds
(9/13) Running minio-java tests ... done in 53 seconds
(10/13) Running minio-js tests ... FAILED in 25 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-41234bfa-f46b-4111-84f4-6d6038e40da0, events: s3:ObjectCreated:*",
  "duration": 20021,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#8f7ff4d7-67a3-47e0-8767-8c78b8ad46d2) |
<a name="8f7ff4d7-67a3-47e0-8767-8c78b8ad46d2"></a>
### 6910-749108d/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32529
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c3db619f6154:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 43 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 31 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 45 seconds
(8/13) Running minio-go tests ... done in 22 seconds
(9/13) Running minio-java tests ... done in 1 minutes and 36 seconds
(10/13) Running minio-js tests ... FAILED in 27 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-d4acfebb-0cc9-48ee-9325-a9280fe97d1e, events: s3:ObjectCreated:*",
  "duration": 20049,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by sinhaashish** [comment]
population.csv
> 
> ```
> ~ mc sql -e "select * from s3object" myminio/sjm-airlines/population.csv | jq .
> {
>   "PopTotal": "1291.622",
>   "PopFemale": "661.578",
>   "PopMale": "630.044",
>   "AgeGrpSpan": "5",
>   "AgeGrpStart": "0",
>   "AgeGrp": "0-4",
>   "MidPeriod": "1950.5",
>   "Time": "1950",
>   "Variant": "High",
>   "VarID": "3",
>   "Location": "Afghanistan",
>   "LocID": "4"
> }
> ...
> ```
The Json output is just in the reverse order as the input csv, I mean the first coulmn in csv  is the last tag in json , the second column of csv as sencond last tag in json and so on . Can we maintain the sequence in output json is possible.

**Comment by sinhaashish** [comment]
For JSON 1:
```
{"Id": 2, "Name": "Ben Mostafa", "Address": "Las Vegas", "Car": "Mustang","Pin": "560048", "Phone": "123456789","UID": "8876"}
{"Id": 3, "Name": "Rohan Wood", "Address": "Wooddon", "Car": "VW"}
{"Id": 4, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep" }
{"Id": 1, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep"}
```
The CSV  output is correct as shown below 
```
2,Ben Mostafa,Las Vegas,Mustang,560048,123456789,8876
3,Rohan Wood,Wooddon,VW,,,
4,Anshu,Templestowe,Jeep,,,
1,Anshu,Templestowe,Jeep,,,
```
But for Json of the form shown below; where the second Json element has more tags than the first JSON element , the output is not correct.
```
{"Id": 1, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep"}
{"Id": 2, "Name": "Ben Mostafa", "Address": "Las Vegas", "Car": "Mustang","Pin": "560048", "Phone": "123456789","UID": "8876"}
{"Id": 3, "Name": "Rohan Wood", "Address": "Wooddon", "Car": "VW"}
{"Id": 4, "Name": "Anshu", "Address": "Templestowe", "Car": "Jeep" }
```
Output does not have the `PIN` , `Phone` and `UID` values
```
1,Anshu,Templestowe,Jeep
2,Ben Mostafa,Las Vegas,Mustang
3,Rohan Wood,Wooddon,VW
4,Anshu,Templestowe,Jeep

```

**Comment by harshavardhana** [comment]
Done addressed both situations @sinhaashish 

- CSV column order preserved in JSON output
- JSON multi-line variable keys handled for CSV output

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#b94dd11e-e0f8-4072-846e-c96d7ec119c6) |
<a name="b94dd11e-e0f8-4072-846e-c96d7ec119c6"></a>
### 6910-9dda20e/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32147
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c9ef9017772d:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 42 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 8 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... done in 16 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 41 seconds
(8/13) Running minio-go tests ... done in 49 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 51 seconds
(10/13) Running minio-js tests ... FAILED in 31 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-5fb21d61-5366-4292-9dc6-29a2fb6ef3a2, events: s3:ObjectCreated:*",
  "duration": 20045,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#7b02061d-9b56-43ad-a12b-10bbb99a2ea5) |
<a name="7b02061d-9b56-43ad-a12b-10bbb99a2ea5"></a>
### 6910-4f1dde1/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31451
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 464293aa06bf:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 38 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 1 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 8 seconds
(6/13) Running mc tests ... done in 31 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 38 seconds
(8/13) Running minio-go tests ... done in 28 seconds
(9/13) Running minio-java tests ... done in 1 minutes and 52 seconds
(10/13) Running minio-js tests ... FAILED in 33 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-e71b5eae-a1ef-4c99-adf6-aed804a94d6c, events: s3:ObjectCreated:*",
  "duration": 20043,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
ping @sinhaashish @donatello can you review this again?

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#5094863a-941d-46b5-a217-f932c1a2e1ca) |
<a name="5094863a-941d-46b5-a217-f932c1a2e1ca"></a>
### 6910-5aef684/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30838
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cd2ec6df4c26:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 40 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 6 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... done in 25 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 51 seconds
(8/13) Running minio-go tests ... done in 48 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 51 seconds
(10/13) Running minio-js tests ... FAILED in 30 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-78aa7b2c-6e73-4c68-8f25-17aee05aad28, events: s3:ObjectCreated:*",
  "duration": 20045,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:488:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:377:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#67ec8957-fe8a-4dc5-bfe9-99d89ecaf280) |
<a name="67ec8957-fe8a-4dc5-bfe9-99d89ecaf280"></a>
### 6910-204cc73/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32299
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 694f2fc55801:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 43 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 35 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 43 seconds
(8/13) Running minio-go tests ... done in 51 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 9 seconds
(10/13) Running minio-js tests ... FAILED in 31 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-d2ccc9d1-127d-481c-84af-05e9d02aef52, events: s3:ObjectCreated:*",
  "duration": 20016,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:488:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:377:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
@sinhaashish @donatello is there anything pending on this PR, can we move ahead with reviews?

**Review by sinhaashish — APPROVED** [review]
Tested & LGTM

File: pkg/s3select/format/csv/csv.go [hunk 1]
```diff
@@ -173,8 +182,8 @@ func (reader *cinput) Read() ([]byte, error) {
 	if dec != nil {
 		var data []byte
 		var err error
-		for i, value := range dec {
-			data, err = sjson.SetBytes(data, reader.header[i], value)
+		for i := len(dec) - 1; i >= 0; i-- {
```
Comment (donatello) [inline pkg/s3select/format/csv/csv.go:190]: Is it being done in reverse for performance?

Comment (harshavardhana) [reply pkg/s3select/format/csv/csv.go:190]: No it is done to keep the order.

**Review by donatello — APPROVED** [review]

**Pull request merged**